packages/backend/src/cache/middleware.ts

```diff
@@ -0,0 +1,118 @@
+import isEqual from 'lodash/isEqual'
+import { Logger } from 'pino'
+import { CacheDataStore } from './data-stores'
+
+interface CachedRequest {
+  requestResult: unknown
+  requestParams: Record<string, unknown>
+  operationName: string
+}
+
+type Request = () => Promise<unknown>
+
+interface CacheMiddlewareArgs {
+  deps: { logger: Logger; dataStore: CacheDataStore }
+  idempotencyKey?: string
+  request: Request
+  requestParams: Record<string, unknown>
+  operationName: string
+  handleParamMismatch: () => never
+}
+
+export async function cacheMiddleware(
+  args: CacheMiddlewareArgs
+): ReturnType<Request> {
+  const {
+    deps: { dataStore },
+    idempotencyKey,
+    request
+  } = args
+  if (!idempotencyKey) {
+    return request()
+  }
+
+  const cachedRequestAsString = await dataStore.get(idempotencyKey)
+
+  return cachedRequestAsString
+    ? handleCacheHit({ ...args, idempotencyKey }, cachedRequestAsString)
+    : handleCacheMiss({ ...args, idempotencyKey })
```

Comment:
```suggestion
    ? handleCacheHit(args, cachedRequestAsString)
    : handleCacheMiss(args)
```
Why do you have to de- and re-construct the args? `idempotencyKey` is a property of `args`, so wouldn't this work?